.. Immcantation: An Integrated Framework for Adaptive Immune Receptor Repertoire Analysis
.. keywords:  rep-seq, immuno-seq, repertoire sequencing, BCR, TCR, Ig, IgOme, AIRR, adaptive immunity,
    somatic hypermutation, vdj-seq, AbSeq, AbPair, iRepertoire, VDJ, Immunoglobulin

.. toctree::
    :maxdepth: 1
    :hidden:

    Welcome <self>

.. _Welcome:

Welcome to the Immcantation Portal!
==========================================================================================

Advances in high-throughput sequencing technologies now allow for large-scale
characterization of B cell receptor (BCR) and T cell receptor (TCR) repertoires. The high
germline and somatic diversity of the adaptive immune receptor repertoire (AIRR) presents
challenges for biologically meaningful analysis - requiring the development of specialized
computational methods.

The Immcantation tool suites provide a start-to-finish analytical ecosystem for
high-throughput AIRR-seq datasets. Beginning from raw reads, Python and R packages are
provided for pre-processing, population structure determination, and repertoire analysis.

**Click on the relevant section of the image below for more details.**

.. raw:: html

    <object data="_static/overview.svg" type="image/svg+xml"></object>
    <br><br>

**For an overview of AIRR-seq analysis, check out:**

    **Yaari and Kleinstein.**
    Practical guidelines for B-cell receptor repertoire sequencing analysis.
    *Genome Medicine. 7, 121 (2015).*
    `doi\:10.1186/s13073-015-0243-2 <http://doi.org/10.1186/s13073-015-0243-2>`__

   **AIRR-seq analysis tutorial**
   `Slides<http://clip.med.yale.edu/immcantation/AIRR-seq_workshop_2017-06-19/AIRR-seq_workshop_slides_2017-06-19.pptx>` 
    and `data<http://clip.med.yale.edu/immcantation/AIRR-seq_workshop_2017-06-19/AIRR-seq_workshop_example.tar.gz>` 
    from the tutorial presented by Jason Vander Heiden at the 
   `Stochasticity and Control of Immune Repertoires Group Conference
   <http://ias.huji.ac.il/content/stochasticity-and-control-dynamics-and-diversity-immune-repertories>` 
   (Israel, June 18 - June 23, 2017).

.. _Cite:

How to Cite
==========================================================================================

To cite the pRESTO software package in publications please use:

    **Vander Heiden JA\*, Yaari G\*, Uduman M, Stern JNH, O'Connor KC, Hafler DA, Vigneault F, Kleinstein SH.**
    pRESTO\: a toolkit for processing high-throughput sequencing raw reads of lymphocyte receptor repertoires.
    *Bioinformatics 30, 1930-2 (2014).*
    `doi\:10.1093/bioinformatics/btu138 <http://doi.org/10.1093/bioinformatics/btu138>`__

To cite the Change-O, Alakazam, SHazaM and TIgGER software package in publications please use:

    **Gupta NT\*, Vander Heiden JA\*, Uduman M, Gadala-Maria D, Yaari G, Kleinstein SH.**
    Change-O\: a toolkit for analyzing large-scale B cell immunoglobulin repertoire sequencing data.
    *Bioinformatics 31, 3356-8 (2015).*
    `doi\:10.1093/bioinformatics/btv359 <http://doi.org/10.1093/bioinformatics/btv359>`__

Additional citations for specific methods within Alakazam, SHazaM and TIgGER may be determined
using the ``citation()`` function witin R.

.. _Contact:

Contact Information
==========================================================================================

If you have questions you can email
`Steven Kleinstein <mailto:steven.kleinstein@yale.edu>`__.

For additional computational immunology software from the Kleinstein Lab see our
`website <http://medicine.yale.edu/lab/kleinstein/software/>`__.

.. Publications that use Immcantation tools:

.. Component TOC trees

.. toctree::
    :maxdepth: 1
    :caption: pRESTO
    :hidden:

    Documentation <http://presto.readthedocs.io>
    Download Releases <http://pypi.python.org/pypi/presto>
    Source Code <http://bitbucket.org/kleinstein/presto>

.. toctree::
    :maxdepth: 1
    :caption: Change-O
    :hidden:

    Documentation <http://changeo.readthedocs.io>
    Download Releases <http://pypi.python.org/pypi/changeo>
    Source Code <http://bitbucket.org/kleinstein/changeo>

.. toctree::
    :maxdepth: 1
    :caption: Alakazam
    :hidden:

    Documentation <http://alakazam.readthedocs.io>
    Download Releases <http://cran.rstudio.com/web/packages/alakazam>
    Source Code <http://bitbucket.org/kleinstein/alakazam>

.. toctree::
    :maxdepth: 1
    :caption: SHazaM
    :hidden:

    Documentation <http://shazam.readthedocs.io>
    Download Releases <http://cran.rstudio.com/web/packages/shazam>
    Source Code <http://bitbucket.org/kleinstein/shazam>

.. toctree::
    :maxdepth: 1
    :caption: TIgGER
    :hidden:

    Documentation <http://tigger.readthedocs.org>
    Download Releases <http://bitbucket.org/kleinstein/tigger/downloads>
    Source Code <http://bitbucket.org/kleinstein/tigger>
